src/main/java/org/snomed/snowstorm/rest/ConceptController.java

```diff
@@ -554,12 +587,50 @@ public Collection<ConceptMini> findConceptParents(@PathVariable String branch,
 	public Collection<?> findConceptAncestors(@PathVariable String branch,
 			@PathVariable String conceptId,
 			@RequestParam(defaultValue = "inferred") Relationship.CharacteristicType form,
-			@RequestHeader(value = "Accept-Language", defaultValue = Config.DEFAULT_ACCEPT_LANG_HEADER) String acceptLanguageHeader) {
+			@RequestHeader(value = "Accept-Language", defaultValue = Config.DEFAULT_ACCEPT_LANG_HEADER) String acceptLanguageHeader) throws ServiceException {
 
 		branch = BranchPathUriUtil.decodePath(branch);
 		return findConceptsWithECL(">" + conceptId, form == Relationship.CharacteristicType.stated, branch, acceptLanguageHeader, 0, LARGE_PAGE.getPageSize()).getItems();
 	}
 
+	@GetMapping(value = "/browser/{branch}/concepts/ancestor-paths")
+	@JsonView(value = View.Component.class)
+	public Collection<ConceptMini> findConceptAncestorPaths(@PathVariable String branch,
+			@RequestParam(required = false) List<Long> conceptIds,		
+			@RequestParam(defaultValue = "inferred") Relationship.CharacteristicType form,
+			@RequestHeader(value = "Accept-Language", defaultValue = Config.DEFAULT_ACCEPT_LANG_HEADER) String acceptLanguageHeader) throws ServiceException {
+
+		branch = BranchPathUriUtil.decodePath(branch);	
+		
+		Map<String, ConceptMini> conceptMiniMap = conceptService.findConceptMinis(branch, conceptIds, ControllerHelper.parseAcceptLanguageHeaderWithDefaultFallback(acceptLanguageHeader)).getResultsMap();
+		
+		// For each concept, lookup a single ancestor-path from it to the top-level concept, and add the path to the result output.
+		Collection<ConceptMini> conceptsWithAncestorPaths = new ArrayList<>();
+		
+		for(final String conceptId : conceptMiniMap.keySet()) {
+			ArrayList<ConceptMini> ancestorPath = ancestorPathHelper(branch, form, conceptId, new ArrayList(), acceptLanguageHeader);
+			conceptMiniMap.get(conceptId).addExtraField("descriptions", conceptMiniMap.get(conceptId).getActiveDescriptions());
+			conceptMiniMap.get(conceptId).addExtraField("ancestorPath", ancestorPath);
+			conceptsWithAncestorPaths.add(conceptMiniMap.get(conceptId));
+		}
+		
+		return conceptsWithAncestorPaths;
+	}	
+	
+	private ArrayList<ConceptMini> ancestorPathHelper(String branch, Relationship.CharacteristicType form, String conceptId, ArrayList<ConceptMini> pathSoFar, String acceptLanguageHeader) throws ServiceException {
+		Collection<ConceptMini> conceptParents = findConceptParents(branch, conceptId, form, false, acceptLanguageHeader);
```

Comment:
**Performance Observation**
Currently the method iterates through concepts and for each concept hits Elasticsearch in a recursive method to load ancestor paths. 
For 10 concepts with an average depth in the hierarchy of 7 there will be at least 70 Elasticsearch calls. I assume this is very slow and gets progressively slower with more concepts.

**Optimisation Suggestion**
We could load the ancestor paths of around 500 concepts at once using just 2 Elasticsearch calls
- The first query can load the unordered set of ancestors for all concepts at once
  - Use a query against the QueryConcept type with `must(term("stated", form == stated)).must(terms("conceptIdL, allConcepts))`
  - Gather concept ids from `ancestors` field
- The second query can load the direct parents of all the relevant concepts
  - Query against QueryConcept type again with `must(term("stated", form == stated)).must(terms("conceptIdL, allConceptsPlusAncestors))`
  - From this a hierarchy can be created in memory. The ancestor paths can be extracted from this.